cryptography/src/bls12381/primitives/group.rs

```diff
@@ -493,6 +518,64 @@ impl Point for G1 {
             );
         }
     }
+
+    /// Performs multi-scalar multiplication (MSM) on G1 points using Pippenger's algorithm.
+    /// Computes `sum(scalars[i] * points[i])`.
+    ///
+    /// Filters out pairs where the point is the identity element (infinity).
+    /// Returns an error if the lengths of the input slices mismatch.
+    fn msm(points: &[Self], scalars: &[Scalar]) -> Self {
+        // Assert input validity
+        assert_eq!(points.len(), scalars.len(), "mismatched lengths");
+
+        // Prepare points (affine) and scalars (raw blst_scalar)
+        let mut points_filtered = Vec::with_capacity(points.len());
+        let mut scalars_filtered = Vec::with_capacity(scalars.len());
+        for (point, scalar) in points.iter().zip(scalars.iter()) {
+            // `blst` does not filter out infinity, so we must ensure it is impossible.
+            //
+            // Sources:
+            // * https://github.com/supranational/blst/blob/cbc7e166a10d7286b91a3a7bea341e708962db13/src/multi_scalar.c#L10-L12
+            // * https://github.com/MystenLabs/fastcrypto/blob/0acf0ff1a163c60e0dec1e16e4fbad4a4cf853bd/fastcrypto/src/groups/bls12381.rs#L160-L194
+            if *point == G1::zero() || scalar == &Scalar::zero() {
+                continue;
+            }
+
+            // Add to filtered vectors
+            points_filtered.push(point.as_blst_p1_affine());
+            scalars_filtered.push(scalar.as_blst_scalar());
+        }
+
+        // If all points were filtered, return zero.
+        if points_filtered.is_empty() {
+            return G1::zero();
+        }
+
+        // Create vectors of pointers for the blst API.
+        // These vectors hold pointers *to* the elements in the filtered vectors above.
+        let points: Vec<*const blst_p1_affine> =
+            points_filtered.iter().map(|p| p as *const _).collect();
+        let scalars: Vec<*const u8> = scalars_filtered.iter().map(|s| s.b.as_ptr()).collect();
```

Comment:
Verify that this is meant to be a vector of pointers to scalars and not a vector of scalars (it seems like the tests would've failed if not).

Reply:
This is correct AFAICT: https://github.com/supranational/blst/blob/cbc7e166a10d7286b91a3a7bea341e708962db13/bindings/rust/src/bindings.rs#L833-L842

The bindings look different approach but are ultimately the same thing: https://github.com/supranational/blst/blob/cbc7e166a10d7286b91a3a7bea341e708962db13/bindings/rust/src/pippenger.rs#L144-L162